Add support for Web Extensions permissions.request() API.

kiaraarose opened the pull request:
#### 65e4c3db6cba94a2f2d131433033cf3e15f30c1a
<pre>
Add support for Web Extensions permissions.request() API.
<a href="https://bugs.webkit.org/show_bug.cgi?id=250135">https://bugs.webkit.org/show_bug.cgi?id=250135</a>
Reviewed by Timothy Hatcher.

This patch implements the permissions.request() API. This includes delegating the action of requesting
permissions from the users to the browsers. Once that action has been fulfilled and the user grants
all requested permissions, we add them to the extension&apos;s granted permissions and match pattern sets.

* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::toAPI): Helper method to convert from HashSet&lt;String&gt; to NSSet&lt;String&gt;.
(WebKit::toAPIArray):Helper method to convert from HashSet&lt;String&gt; to NSArray&lt;String&gt;.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerDelegate.h:
Define the delegate methods used to prompt users to grant permissions.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIEventCocoa.mm:
Remove unnecessary file import.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
(WebKit::WebExtensionContext::permissionsContains):

(WebKit::WebExtensionContext::permissionsRequest):
Call delegate methods to determine if permissions should be granted.

(WebKit::WebExtensionContext::permissionsRemove):

(WebKit::WebExtensionContext::firePermissionsEventListenerIfNecessary):
Fire permissions onAdded or onRemoved events if permissions are granted or removed.

(WebKit::WebExtensionContext::parseMatchPatterns): Deleted.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::postAsyncNotification):
Call firePermissionsEventListenerIfNecessary whenever a notification is posted for a change in an
extension&apos;s permissions.

(WebKit::WebExtensionContext::hasPermissions):
(WebKit::WebExtensionContext::wakeUpBackgroundContentIfNecessaryToFireEvents):
`fireEvents` isn’t an accurate name for this method since it doesn’t actually fire any events.
We should rename this to better describe what’s happening in this method
(WebKit::WebExtensionContext::fireEvents): Deleted.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::didStartProvisionalLoadForFrame):
(WebKit::WebExtensionController::didCommitLoadForFrame):
(WebKit::WebExtensionController::didFinishLoadForFrame):
(WebKit::WebExtensionController::didFailLoadForFrame):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMatchPatternCocoa.mm:
(WebKit::WebExtensionMatchPattern::matchPatternStrings):
(WebKit::WebExtensionMatchPattern::matchPatternsForOrigins):

* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
Validate permissions details for these APIs.
(WebKit::WebExtensionAPIPermissions::contains):
(WebKit::WebExtensionAPIPermissions::request):
(WebKit::WebExtensionAPIPermissions::remove):
(WebKit::WebExtensionAPIPermissions::verifyRequestedPermissions):

(WebKit::WebExtensionAPIPermissions::onAdded):
(WebKit::WebExtensionAPIPermissions::onRemoved):
Pass in the WebExtensionEventListenerType when creating these events.

* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
(WebKit::WebExtensionContextProxy::dispatchPermissionsEvent):
Enumerate through the namespace objects and invoke the listeners for the event listener.

* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIPermissions.mm: Added.
(TestWebKitAPI::runScriptWithUserGesture):
(TestWebKitAPI::TEST):
Add WebKit tests for Permissions API.

* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.h:
Copied from Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerDelegate.h.

* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.mm:
Copied from Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in.
(-[TestWebExtensionsDelegate
webExtensionController:promptForPermissions:inTab:forExtensionContext:completionHandler:]):
(-[TestWebExtensionsDelegate
webExtensionController:promptForPermissionMatchPatterns:inTab:forExtensionContext:completionHandler:]):

Canonical link: <a href="https://commits.webkit.org/260370@main">https://commits.webkit.org/260370@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/be41e1ef323d92a4e3caa09beeb6787273b8d5d4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108038 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17098 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40913 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117166 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116498 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111926 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18470 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8408 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100236 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113804 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/13939 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97154 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41861 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95847 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28792 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83530 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9999 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30140 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10717 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7041 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16149 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49730 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12296 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3900 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -70,12 +72,74 @@
+            [delegate webExtensionController:extensionController()->wrapper() promptForPermissionMatchPatterns:patterns inTab:nil forExtensionContext:wrapper() completionHandler:originsCompletionHandler];
+        } else
+            completionHandler(false);
+
```
Remove extra newline

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -70,12 +72,73 @@
+    };
+
+    auto permissionCompletionHandler = ^(NSSet<_WKWebExtensionPermission> *allowedPermissions) {
+        // User did not grant all requested permissions.
```
Figured we should return false if not all the permissions were granted. I can test in Chrome to see what the behavior is there.

kiaraarose replied:
Unless the expected behavior is that the user either grants all the permissions or none at all?

xeenon replied:
The client could grant some, or all. I'm not sure what the right thing to return would be. Hmm

kiaraarose replied:
Tim and I decided to make the API require all requested permissions to be granted in order for it to return `true`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -70,12 +72,73 @@
 
-    // FIXME: <https://webkit.org/b/250135> Call delegate method to prompt user for access.
+    // There shouldn't be any unsupported permissions, since they got reported as an error before this.
+    ASSERT(!permissions.isSubset(extension().supportedPermissions()));
```
Should this be: `ASSERT(permissions.isSubset(extension().supportedPermissions()));`?

kiaraarose replied:
Yes! I'll fix this.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -70,12 +72,73 @@
+        if ([delegate respondsToSelector:@selector(webExtensionController:promptForPermissionMatchPatterns:inTab:forExtensionContext:completionHandler:)]) {
+            NSMutableSet *patterns = [NSMutableSet set];
+            for (auto& matchPattern : matchPatterns)
+                [patterns addObject:[_WKWebExtensionMatchPattern matchPatternWithString:static_cast<NSString *>(matchPattern.get().string())]];
```
This should work: `[patterns addObject:matchPattern->wrapper()];`

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -497,6 +497,9 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
         return;
 
+    MatchPatternSet matchPatterns;
+    firePermissionsEventListenerIfNecessary(notificationName, permissions, matchPatterns);
```
Is there a better way to do this? Like pass in an empty set in the function call? If not, I can split this function into two, one for match patterns and the other for permissions if it seems weird to create these sets but not really use them.

xeenon replied:
I think just putting `{ }` in the call instead of `matchPatterns ` will work.

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -742,7 +748,7 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
         bool matchFound = false;
         for (auto& grantedPattern : currentPermissionMatchPatterns()) {
-            if (pattern->matchesPattern(grantedPattern, { WebExtensionMatchPattern::Options::IgnorePaths })) {
+            if (grantedPattern->matchesPattern(pattern, { WebExtensionMatchPattern::Options::IgnorePaths })) {
```
I'm not 100% convinced switching this is correct. will confirm in the morning when I re-test this.

xeenon replied:
Yeah I think this change is correct.

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -123,34 +141,33 @@
-            if (requestedPattern->matchesPattern(allowedPattern, { WebExtensionMatchPattern::Options::IgnorePaths })) {
-                requestingAllowedOrigin = true;
-                break;
+            if (bool requestingDisallowedOrigin = !requestedPattern->matchesPattern(allowedPattern, { WebExtensionMatchPattern::Options::IgnorePaths })) {
```
I think this method needs a bit more work. I believe I should be checking the optional match patterns here as well and not return an error message if the requested pattern isn't in the manifest. will confirm in the morning when I re-test this.

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -68,14 +73,82 @@
+
+    if ([delegate respondsToSelector:@selector(webExtensionController:promptForPermissions:inTab:forExtensionContext:completionHandler:)]) {
+        NSMutableSet *permissionsToRequest = [NSMutableSet set];
+        for (auto& permission : permissions)
```
I think I might add a helper method in Cocoa helpers that translates a NSSet to HashSet and vice versa

xeenon replied:
Yeah, for the `String` case CocoaHelpers.mm/h would be the a fine place to put that. Or a new HashSetCocoa.h similar to VectorCocoa.h and `createNSArray()` and `makeVector()`.

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -107,6 +110,21 @@
+        return;
+    }
+
+    if (!verifyRequestedPermissions(permissions, matchPatterns, @"permissions.remove()", &errorMessage)) {
```
Will change this to report an error regardless now that `remove()` returns an error if requesting to remove a permission not defined in the manifest.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -68,14 +73,82 @@
+            requestedAllHostsPattern = true;
+            break;
+        }
+    }
```
`WebExtensionMatchPattern::patternsMatchAllHosts()` can be used here.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -68,14 +73,82 @@
+        if ([delegate respondsToSelector:@selector(webExtensionController:promptForPermissionMatchPatterns:inTab:forExtensionContext:completionHandler:)]) {
+            NSMutableSet *matchPatternsToRequest = [NSMutableSet set];
+            for (auto& matchPattern : matchPatterns)
+                [matchPatternsToRequest addObject:matchPattern->wrapper()];
```
A couple helper static methods or freestanding functions in WebExtensionMatchPattern.h would be good for this. See the `toImpl()` and `toAPI()` functions in _WKWebExtensionContext.mm that I have for dictionary versions of these. You could add NSArray versions of `toImpl()` and `toAPI()` functions in WebExtensionMatchPattern.h.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -68,14 +73,82 @@
+
+        grantPermissionMatchPatterns(WTFMove(grantedPatterns));
+        grantPermissions(WTFMove(grantedPermissions));
+        completionHandler(true);
```
Missing `firePermissionsEventListenerIfNecessary(…)`?

kiaraarose replied:
I put the call to `firePermissionsEventListenerIfNecessary()` in the `postAsyncNotification()` functions so that I would only have to call them in one place.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -99,14 +172,36 @@
+
+    HashSet<String> origins;
+    for (auto& matchPattern : matchPatterns)
+        origins.add(matchPattern.get().string());
```
Another good spot for a helper.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -99,14 +172,36 @@
 
+void WebExtensionContext::firePermissionsEventListenerIfNecessary(NSString *notificationName, PermissionsSet& permissions, MatchPatternSet& matchPatterns)
+{
+    if (!([notificationName isEqualToString:_WKWebExtensionContextPermissionsWereGrantedNotification] || [notificationName isEqualToString:_WKWebExtensionContextGrantedPermissionsWereRemovedNotification]))
```
This should likely be an `ASSERT` of the type event types.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -99,14 +172,36 @@
     }
 }
 
+void WebExtensionContext::firePermissionsEventListenerIfNecessary(NSString *notificationName, PermissionsSet& permissions, MatchPatternSet& matchPatterns)
```
It feels odd to use the `NSString *notificationName ` here and not have it be used other than to know what event to fire.

This would be better done with `WebExtensionEventListenerType`, and just assert the two event types.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -99,14 +172,36 @@
+        origins.add(matchPattern.get().string());
+
+    if ([notificationName isEqualToString:_WKWebExtensionContextPermissionsWereGrantedNotification]) {
+        auto listenerTypes = WebExtensionContext::EventListenerTypeSet { WebExtensionEventListenerType::PermissionsOnAdded };
```
Can move out of the `if` and use the passed in `eventType`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -99,14 +172,36 @@
+            sendToProcessesForEvent(WebExtensionEventListenerType::PermissionsOnAdded, Messages::WebExtensionContextProxy::DispatchPermissionsEvent("onAdded"_s, permissions, origins));
+        });
+    } else {
+        auto listenerTypes = WebExtensionContext::EventListenerTypeSet { WebExtensionEventListenerType::PermissionsOnRemoved };
```
Ditto.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -507,6 +510,9 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
         return;
 
+    PermissionsSet permissions;
+    firePermissionsEventListenerIfNecessary(notificationName, permissions, matchPatterns);
```
Ditto on using `{ }`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -99,14 +172,36 @@
+
+    if ([notificationName isEqualToString:_WKWebExtensionContextPermissionsWereGrantedNotification]) {
+        auto listenerTypes = WebExtensionContext::EventListenerTypeSet { WebExtensionEventListenerType::PermissionsOnAdded };
+        wakeUpBackgroundContentIfNecessaryToFireEvents(listenerTypes, [permissions, origins, this] {
```
You should change this to `[&permissions, &origins, this]` to avoid copying the sets. Or just do `[&]` to let the compiler reference what you use in the lambda.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -99,14 +172,36 @@
+        });
+    } else {
+        auto listenerTypes = WebExtensionContext::EventListenerTypeSet { WebExtensionEventListenerType::PermissionsOnRemoved };
+        wakeUpBackgroundContentIfNecessaryToFireEvents(listenerTypes, [permissions, origins, this] {
```
Ditto on `[&]`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -247,7 +247,7 @@
             continue;
 
-        context->fireEvents(listenerTypes, [context, pageID, frameID, targetURL] {
+        context->wakeUpBackgroundContentIfNecessaryToFireEvents(listenerTypes, [context, pageID, frameID, targetURL] {
```
Should change these to `[&]` too.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -262,7 +262,7 @@
             continue;
 
-        context->fireEvents(listenerTypes, [context, pageID, frameID, frameURL] {
+        context->wakeUpBackgroundContentIfNecessaryToFireEvents(listenerTypes, [context, pageID, frameID, frameURL] {
```
Ditto on `[&]`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -278,7 +278,7 @@
             continue;
 
-        context->fireEvents(listenerTypes, [context, pageID, frameID, frameURL] {
+        context->wakeUpBackgroundContentIfNecessaryToFireEvents(listenerTypes, [context, pageID, frameID, frameURL] {
```
Ditto on `[&]`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -293,7 +293,7 @@
             continue;
 
-        context->fireEvents(listenerTypes, [context, pageID, frameID, frameURL] {
+        context->wakeUpBackgroundContentIfNecessaryToFireEvents(listenerTypes, [context, pageID, frameID, frameURL] {
```
Ditto on `[&]`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -63,7 +63,7 @@
     NSString *errorMessage;
-    if (!validatePermissionsDetails(permissions, origins, matchPatterns, @"contains()", &errorMessage)) {
+    if (!validatePermissionsDetails(permissions, origins, matchPatterns, @"permissions.contains()", &errorMessage)) {
         callback->reportError(errorMessage);
```
This should not use `reportError()`. You should be setting the out string `NSString **` of this method to report the exception. `reportError()` is only used for async errors in completion handlers. Ditto for the rest of the methods that use `validatePermissionsDetails()` like this. You van just pass the `errorMessage` pointer and drop the local `NSString *errorMessage`.

kiaraarose replied:
Spoke with Tim about this offline. Chrome doesn't throw an exception in this case, they report an error using the callback. We should do the same. Will add a comment to explain this.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -83,14 +83,17 @@
 
-    // FIXME: <https://webkit.org/b/250135> Check to see if the request is being made from a user gesture.
+    if (!WebCore::UserGestureIndicator::processingUserGesture()) {
+        callback->reportError(@"permissions.request() must be called during a user gesture.");
```
Ditto.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -83,14 +83,17 @@
 
-    if (!validatePermissionsDetails(permissions, origins, matchPatterns, @"request()", &errorMessage)) {
+    if (!validatePermissionsDetails(permissions, origins, matchPatterns, @"permissions.request()", &errorMessage)) {
         callback->reportError(errorMessage);
```
Ditto.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -83,14 +83,17 @@
 
-    if (!verifyRequestedPermissions(permissions, matchPatterns, @"request()", &errorMessage)) {
+    if (!verifyRequestedPermissions(permissions, matchPatterns, @"permissions.request()", &errorMessage)) {
         callback->reportError(errorMessage);
```
Ditto.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -107,6 +110,21 @@
+    NSString *errorMessage;
+    WebExtension::MatchPatternSet matchPatterns;
+    if (!validatePermissionsDetails(permissions, origins, matchPatterns, @"permissions.remove()", &errorMessage)) {
+        callback->reportError(errorMessage);
```
Ditto.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -107,6 +110,21 @@
+        if (errorMessage)
+            callback->reportError(errorMessage);
+        else
+            callback->call(@NO);
```
Ditto on how `errorMessage` should be handled. Would just return early `if (errorMessage)`. Will `verifyRequestedPermissions` early return `false` and not set an `errorMessage`?

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -123,32 +141,46 @@
-    bool requestingPermissionsNotDefinedInManifest = permissions.size() && !permissions.isSubset(allowedPermissions);
+    bool requestingPermissionsNotDeclaredInManifest = (permissions.size() && !permissions.isSubset(allowedPermissions)) || (matchPatterns.size() && !allowedHostPermissions.size());
+    if (requestingPermissionsNotDeclaredInManifest) {
+        *outExceptionString = [callingAPIName isEqualToString:@"permissions.remove()"] ? @"Only permissions specified in the manifest may be removed." : @"Only permissions specified in the manifest may be requested.";
```
This error is confusing. It says they can be removed, but we error if they try to remove them. Which is it?

I think only optional permissions can be removed. Should verify what Chrome/Firefox allows.

kiaraarose replied:
This is the case where they try to remove or request a permission that's not declared in the manifest. So if they try to remove `scripting` but they never declared it. Chrome throws an error in this case.

xeenon commented on Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
```diff
@@ -41,6 +42,27 @@
 
+// MARK: Permissions support
+
+void WebExtensionContextProxy::dispatchPermissionsEvent(const String& eventType, HashSet<String> permissions, HashSet<String> origins)
```
It would be better to use the `WebExtensionEventListenerType` instead of a String. And `ASSERT` it is only the two types.

xeenon commented on Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
```diff
@@ -41,6 +42,27 @@
+
+    NSMutableArray<NSString *> *originDetails = [NSMutableArray array];
+    for (auto& origin : origins)
+        [originDetails addObject:origin];
```
More good spots for helper methods.

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -60,36 +65,94 @@
-    }
+    MatchPatternSet matchPatterns;
+    WebExtensionMatchPattern::matchPatternsForOrigins(origins, matchPatterns);
+
```
Will remove this newline.

kiaraarose commented on Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
```diff
@@ -77,6 +79,9 @@ class WebExtensionMatchPattern : public API::ObjectImpl<API::Object::Type::WebEx
 
     static void registerCustomURLScheme(String);
 
+    static void matchPatternStrings(const MatchPatternSet&, HashSet<String>&);
```
Open to better name suggestions for these helper methods

xeenon replied:
This can return the result, as `HashSet<String>&&`. Maybe call this `HashSet<String>&& toStrings(const MatchPatternSet&)`.

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -123,32 +141,46 @@
-    if ([callingAPIName isEqualToString:@"remove()"]) {
-        if (bool requestingToRemoveFunctionalPermissions = permissions.size() && permissions.isSubset(allowedPermissions)) {
+    if ([callingAPIName isEqualToString:@"permissions.remove()"]) {
+        if (bool requestingToRemoveFunctionalPermissions = permissions.size() && permissions.intersectionWith(allowedPermissions).size()) {
```
Checking for `isSubset` here wasn't sufficient because it didn't catch the case where the input was a mixed array of optional and required permissions.

kiaraarose commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIPermissions.mm:
```diff
@@ -0,0 +1,204 @@
+    TestWebKitAPI::Util::run(&requestComplete);
+}
+
+TEST(WKWebExtensionAPIPermissions, DenyPermissionsRequestTest)
```
I'm struggling to figure out how to get this test to pass. I confirmed that the call to request returns `false` by adding logs, but for some reason, the test hangs and timeout without a useful error message. Any suggestions on how I can restructure this test?

xeenon replied:
Shouldn't the call to `request()` return `true` since you grant what was requested in the blocks? If it returns `false` that suggests it is returning easy somewhere and/or `requestComplete = true` is not being reached.

The structure of the test looks correct to me.

xeenon replied:
Also can you add `browser.test.assertTrue` around `browser.permissions.request()` like you do for the deny case below?

kiaraarose replied:
Spoke to Tim about this offline. The DenyRequestTest should return false since it's not granting permissions. I added `browser.test.assertTrue` to the AcceptRequestTest. Was able to get the second test to pass.

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.h:
```diff
@@ -78,4 +83,8 @@ NSString *escapeCharactersInString(NSString *, NSString *charactersToEscape);
 WallTime toImpl(NSDate *);
 
+NSSet *toAPI(HashSet<String>&);
+NSSet *toAPI(MatchPatternSet&);
```
This should move to `WebExtensionMatchPattern.h` and `WebExtensionMatchPattern.cpp` outside the `class { }`

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -235,4 +235,31 @@ WallTime toImpl(NSDate *date)
 
+NSSet *toAPI(HashSet<String>& set)
+{
+    NSMutableSet *result = [NSMutableSet set];
```
`[[NSMutableSet alloc] initWithCapacity:set.size()]`

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -235,4 +235,31 @@ WallTime toImpl(NSDate *date)
+    for (auto& element : set)
+        [result addObject:static_cast<NSString *>(element)];
+
+    return result;
```
`return [result copy];` to make it immutable.

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -235,4 +235,31 @@ WallTime toImpl(NSDate *date)
+
+NSSet *toAPI(MatchPatternSet& set)
+{
+    NSMutableSet *result = [NSMutableSet set];
```
Ditto.

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -235,4 +235,31 @@ WallTime toImpl(NSDate *date)
+    for (auto& element : set)
+        [result addObject:element->wrapper()];
+
+    return result;
```
Ditto.

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -235,4 +235,31 @@ WallTime toImpl(NSDate *date)
+
+NSArray *toAPIArray(HashSet<String>& set)
+{
+    NSMutableArray *result = [NSMutableArray array];
```
Ditto.

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -235,4 +235,31 @@ WallTime toImpl(NSDate *date)
+    for (auto& element : set)
+        [result addObject:static_cast<NSString *>(element)];
+
+    return result;
```
Ditto.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
```diff
@@ -77,6 +79,9 @@ class WebExtensionMatchPattern : public API::ObjectImpl<API::Object::Type::WebEx
     static void registerCustomURLScheme(String);
 
+    static void matchPatternStrings(const MatchPatternSet&, HashSet<String>&);
+    static void matchPatternsForOrigins(HashSet<String>&, MatchPatternSet&);
```
Ditto, this can return via `MatchPatternSet&&`. Maybe call this `MatchPatternSet&& fromStrings(HashSet<String>&)`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -99,12 +162,23 @@
+    else if ([notificationName isEqualToString:_WKWebExtensionContextGrantedPermissionsWereRemovedNotification])
+        type = WebExtensionEventListenerType::PermissionsOnRemoved;
+    else
+        return;
```
`ASSERT(type == WebExtensionEventListenerType::PermissionsOnAdded || type == WebExtensionEventListenerType::PermissionsOnRemoved)` would be better at the top.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
```diff
@@ -99,12 +162,23 @@
+    if ([notificationName isEqualToString:_WKWebExtensionContextPermissionsWereGrantedNotification])
+        type = WebExtensionEventListenerType::PermissionsOnAdded;
+    else if ([notificationName isEqualToString:_WKWebExtensionContextGrantedPermissionsWereRemovedNotification])
+        type = WebExtensionEventListenerType::PermissionsOnRemoved;
```
Since the function is now `firePermissionsEventListenerIfNecessary`, I think flipping this would look better at the call sites. So `type` is an argument, and `notificationName` is based on what type is passed.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -107,6 +113,18 @@
+    NSString *errorMessage;
+    WebExtension::MatchPatternSet matchPatterns;
+    if (!validatePermissionsDetails(permissions, origins, matchPatterns, @"permissions.remove()", &errorMessage)) {
+        callback->reportError(errorMessage);
```
Should add the comment: `// Chrome reports this error as callback error and not an exception, so do the same.`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -107,6 +113,18 @@
+    }
+
+    if (!verifyRequestedPermissions(permissions, matchPatterns, @"permissions.remove()", &errorMessage)) {
+        callback->reportError(errorMessage);
```
Ditto on adding the comment.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIPermissions.mm:
```diff
@@ -0,0 +1,204 @@
+{
+    bool callbackComplete = false;
+    RetainPtr<id> evalResult;
+    [backgroundWebView callAsyncJavaScript:script arguments:nil inFrame:nil inContentWorld:WKContentWorld.pageWorld completionHandler:[&] (id result, NSError *error) {
```
`completionHandler:` should take a block (`^`) not a lambda (`[]`). But Maybe in ObjC++ they are one and the same?!

kiaraarose replied:
Maybe? I see a lot of other instances in the codebase where we pass in a lambda here.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIPermissions.mm:
```diff
@@ -0,0 +1,204 @@
+    auto *manifest = @{
+        @"manifest_version": @3,
+        @"permissions": @[ @"alarms", @"activeTab" ],
+        @"optional_permissions": @[ @"webNavigation", @"cookies", @"declarativeNetRequest", @"*://*.apple.com/*" ],
```
I think this is the problem: `@"*://*.apple.com/*"` should be in `optional_host_permission` since this is a v3 extension. It will be ignored in `optional_permissions`.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIPermissions.mm:
```diff
@@ -0,0 +1,204 @@
+        @"permissions": @[ @"alarms", @"activeTab" ],
+        @"optional_permissions": @[ @"webNavigation", @"cookies", @"declarativeNetRequest", @"*://*.apple.com/*" ],
+        @"background": @{ @"scripts": @[ @"background.js" ], @"type": @"module", @"persistent": @NO },
+        @"host_permissions": @[ @"*://*.apple.com/*" ]
```
This should be `optional_host_permission` so it can be requested. Right?

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -54,19 +54,16 @@
     parseDetailsDictionary(details, permissions, origins);
 
-    NSString *errorMessage;
-    if (!validatePermissionsDetails(permissions, origins, matchPatterns, @"contains()", &errorMessage)) {
```
`contains()` doesn't set the error on the callback, only `remove()` and `request()`.

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -235,4 +235,22 @@ WallTime toImpl(NSDate *date)
+
+NSArray *toAPIArray(HashSet<String>& set)
+{
+    NSMutableArray *result = [NSMutableArray array];
```
`[[NSMutableArray alloc] initWithCapacity:set.size()]`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMatchPatternCocoa.mm:
```diff
@@ -396,6 +396,33 @@
 
+HashSet<String> toStrings(const MatchPatternSet& matchPatterns)
+{
+    HashSet<String> stringsToReturn;
```
Add this below: `stringsToReturn.reserveInitialCapacity(matchPatterns.size())` with a blank line after.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMatchPatternCocoa.mm:
```diff
@@ -396,6 +396,33 @@
+
+MatchPatternSet toPatterns(HashSet<String>& domains)
+{
+    MatchPatternSet matchPatterns;
```
Ditto on `reserveInitialCapacity()`.

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -507,6 +512,11 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
         return;
 
+    if ([notificationName isEqualToString:_WKWebExtensionContextPermissionsWereGrantedNotification])
+        firePermissionsEventListenerIfNecessary(WebExtensionEventListenerType::PermissionsOnAdded, { } , matchPatterns);
```
remove extra space after comma

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.h:
```diff
@@ -23,6 +23,7 @@
  * THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+#import <wtf/HashSet.h>
```
I think you need to add `#import <wtf/text/StringHash.h>` to fix the iOS build.

kiaraarose committed:
Add support for Web Extensions permissions.request() API.
https://bugs.webkit.org/show_bug.cgi?id=250135
Reviewed by Timothy Hatcher.

This patch implements the permissions.request() API. This includes delegating the action of requesting
permissions from the users to the browsers. Once that action has been fulfilled and the user grants
all requested permissions, we add them to the extension's granted permissions and match pattern sets.

* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::toAPI): Helper method to convert from HashSet<String> to NSSet<String>.
(WebKit::toAPIArray):Helper method to convert from HashSet<String> to NSArray<String>.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerDelegate.h:
Define the delegate methods used to prompt users to grant permissions.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIEventCocoa.mm:
Remove unnecessary file import.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
(WebKit::WebExtensionContext::permissionsContains):

(WebKit::WebExtensionContext::permissionsRequest):
Call delegate methods to determine if permissions should be granted.

(WebKit::WebExtensionContext::permissionsRemove):

(WebKit::WebExtensionContext::firePermissionsEventListenerIfNecessary):
Fire permissions onAdded or onRemoved events if permissions are granted or removed.

(WebKit::WebExtensionContext::parseMatchPatterns): Deleted.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::postAsyncNotification):
Call firePermissionsEventListenerIfNecessary whenever a notification is posted for a change in an
extension's permissions.

(WebKit::WebExtensionContext::hasPermissions):
(WebKit::WebExtensionContext::wakeUpBackgroundContentIfNecessaryToFireEvents):
`fireEvents` isn’t an accurate name for this method since it doesn’t actually fire any events.
We should rename this to better describe what’s happening in this method
(WebKit::WebExtensionContext::fireEvents): Deleted.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::didStartProvisionalLoadForFrame):
(WebKit::WebExtensionController::didCommitLoadForFrame):
(WebKit::WebExtensionController::didFinishLoadForFrame):
(WebKit::WebExtensionController::didFailLoadForFrame):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMatchPatternCocoa.mm:
(WebKit::WebExtensionMatchPattern::matchPatternStrings):
(WebKit::WebExtensionMatchPattern::matchPatternsForOrigins):

* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
Validate permissions details for these APIs.
(WebKit::WebExtensionAPIPermissions::contains):
(WebKit::WebExtensionAPIPermissions::request):
(WebKit::WebExtensionAPIPermissions::remove):
(WebKit::WebExtensionAPIPermissions::verifyRequestedPermissions):

(WebKit::WebExtensionAPIPermissions::onAdded):
(WebKit::WebExtensionAPIPermissions::onRemoved):
Pass in the WebExtensionEventListenerType when creating these events.

* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
(WebKit::WebExtensionContextProxy::dispatchPermissionsEvent):
Enumerate through the namespace objects and invoke the listeners for the event listener.

* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIPermissions.mm: Added.
(TestWebKitAPI::runScriptWithUserGesture):
(TestWebKitAPI::TEST):
Add WebKit tests for Permissions API.

* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.h:
Copied from Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerDelegate.h.

* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.mm:
Copied from Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in.
(-[TestWebExtensionsDelegate
webExtensionController:promptForPermissions:inTab:forExtensionContext:completionHandler:]):
(-[TestWebExtensionsDelegate
webExtensionController:promptForPermissionMatchPatterns:inTab:forExtensionContext:completionHandler:]):

Canonical link: https://commits.webkit.org/260370@main
